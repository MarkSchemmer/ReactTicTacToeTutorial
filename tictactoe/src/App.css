.App {
  text-align: center;
}

.App-logo {
  animation: App-logo-spin infinite 20s linear;
  height: 40vmin;
}

.App-header {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}


.past-moves {
  width: 200px;
  display:inline-block;
  vertical-align: top;
  margin-right:25px;
  margin-top:45px;
}


.past-moves ul {
  list-style-type : none;
  width: 200px;
}


#board {
  width:400px;
  display:inline-block;
}
