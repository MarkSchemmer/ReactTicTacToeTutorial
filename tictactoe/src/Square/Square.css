

.square span {
    font-size: 60px;
    display: block;
    position: absolute;
    width: 60px;
    margin-left: 24px;
}