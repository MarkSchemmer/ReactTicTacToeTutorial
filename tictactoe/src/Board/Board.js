import React from 'react'
import Square from '../Square/Square'

import * as _ from '../Helpers/calculateWinner'


const Board = (props) => {

    /*
    //JSON.parse(JSON.stringify(this.state.squares))
     is an alternative to cloning the array of objects...
    */


   const { squares, click, xIsNext } = props

       const boardStyles = {
           width:'400px',
           height:'600px',
           margin:'auto'
       }
       let winner = _.calculateWinner(squares)
       let status 
       if(winner === _.X){
         status = <h3>Player X wins</h3>
       } else if (winner === _.O){
         status = <h3>Player O wins</h3> 
       } else if(winner === _.gameOptions.TIE) {
           status = <h3>Tie Game!</h3>
       } else {
           status = <h3>Next Player: <strong> { xIsNext ? 'X' : 'O' } </strong></h3>
       }

        return (
          <React.Fragment>

          <div style={boardStyles} id="board">
                    { status }
                {
                    squares.map((obj,idx) => {
                        return <Square 
                        key={idx} 
                        click={click} 
                        squareNumber={idx} 
                        value={obj} />
                    })
                }
          </div>
          </React.Fragment>
        )
    }

export default Board