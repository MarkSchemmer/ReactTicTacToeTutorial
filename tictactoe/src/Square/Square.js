import React from 'react'
import './Square.css'

const Square = (props) => {

    const { click, squareNumber, value } = props
    const [x,y] = squareNumber

    const squareStyles = {
        width:'110px',
        height:'100px',
        border:'1px solid black',
        display:'inline-block',
        margin:'0',
        marginRight:'2px'
    }

    return (
        <div onClick={() => click(x,y) } style={squareStyles} className="square">
            { value ? <span>{value}</span> : null}
        </div>
    )
}


export default Square