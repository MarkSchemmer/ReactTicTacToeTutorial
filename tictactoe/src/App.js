import React, { Component } from 'react';
import Board from './Board/Board'
import PastMoves from './PastMoves/PastMoves'
import * as _ from './Helpers/calculateWinner'
//import logo from './logo.svg';
import './App.css';

/*


  Adding X,Y coordinates for tic tac toe... 

  the current issue is that I only have a single dimensional array.

  I could hard code the coordinates or I could change the program to work with 
  multidimensions 

  I solved this issue by making squares a multi dimensional array
  then just passing through the right index to the clickSquare function
  and in the calculateWinner function I just flattened the multi dimensional array 


*/

const multiDimTicTacScoreBoard = [...Array(3).keys()]
.map(level => [...Array(3).keys()]
.map(subLevel => ({ _id : Math.floor(Date.now()*Math.random(100)),  
                    val : null, 
                    coor:null, 
                    isWinningSquare:false}) ))

class App extends Component {

  constructor(props){
    super(props)
    this.state = {
      xIsNext:true,
      history :  [ { squares : multiDimTicTacScoreBoard, move:0, coordinates:null } ] ,
      stepNumber:0,
      hasGameToStartAtTop:true 
    }
    this.clickSquare = this.clickSquare.bind(this)
    this.jumpTwo = this.jumpTwo.bind(this)
    this.toggleMoves = this.toggleMoves.bind(this)
    this.winningSquares = this.winningSquares.bind(this)
  }


  toggleMoves () {

    let newHistory = this.state.history.reverse()

    this.setState( (prevState) => ({
      history : newHistory, 
    }))
  }

  clickSquare(x,y) {
    let history = JSON.parse(JSON.stringify(this.state.history))
    let current = history[history.length-1]
    let squares = current.squares.slice().map(subLevels => JSON.parse(JSON.stringify(subLevels)))
    if(_.calculateWinner(squares))
        return 

      if(squares[x][y].val === null){ 
        squares[x][y].val = this.state.xIsNext ? 'X' : 'O'
        squares[x][y].coor = [x,y]
        this.setState( (prevState) => ({
            history : prevState.history.concat( { squares: squares, move: history.length , coordinates:[x,y] } ), 
            stepNumber : history.length, 
            xIsNext : !this.state.xIsNext
        }))
      }
}

togglePlayer() {
   this.setState((prevState) => ({
       xIsNext:!prevState.xIsNext
   }))
}


jumpTwo(step, MOVE=''){
  this.setState({
    stepNumber : step,
    xIsNext : ( step % 2 ) === 0 
  })
}

winningSquares(_squares){
  let history = JSON.parse(JSON.stringify(this.state.history))
  let current = history[history.length-1]
  let squares = current.squares.slice().map(subLevels => JSON.parse(JSON.stringify(subLevels)))
  const [a, b, c] = _squares
  let arr = [a,b,c]

  arr.forEach(ele => {
    ele.isWinningSquare = true
    let x = ele.coor[0], y = ele.coor[1] 
    squares[x][y] = ele 
  })
  current.squares = squares
  history[history.length-1] = current
}



  render() {
    return (
      <div className="App">
        <header className="App-header">
        </header>
        <PastMoves 
          jumpTo={this.jumpTwo} 
          history={this.state.history} 
          toggle={this.toggleMoves}  />
        <Board 
            toggle={this.togglePlayer} 
            xIsNext={this.state.xIsNext}
            click={this.clickSquare} 
            winningSquares={this.winningSquares}
            squares={this.state.history[this.state.stepNumber].squares} />
      </div>
    );
  }
}

export default App;
