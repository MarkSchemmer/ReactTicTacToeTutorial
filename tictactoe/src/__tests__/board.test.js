import React from 'react'
import Board from '../Board/Board'
import Square from '../Square/Square'
import renderer from 'react-test-renderer'
import toJson from 'enzyme-to-json'
import Enzyme, { shallow, mount, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import App from '../App'

Enzyme.configure({ adapter: new Adapter() });

let testIds = [
    [1,2,3],
    [4,5,6],
    [8,9,10]
]

const multiDimTicTacScoreBoard = [...Array(3).keys()]
.map((level,levelIdx) => [...Array(3).keys()]
.map((subLevel, subLevelIdx) => ({ _id : testIds[levelIdx][subLevelIdx],  
                    val : null, 
                    coor:null, 
                    isWinningSquare:false}) ))


describe('Board', () => {
    let wrapper 
    beforeAll(() => {
        wrapper = shallow( <Board
            click={() => {}}
            winningSquares={() => {}}
            xIsNext={true}
            toggle={() => {}} 
            squares={multiDimTicTacScoreBoard} /> )
    })
    it(' has board div? ', () => {
        expect(wrapper.find('#board').length).toBe(1)
    })

    it(' has h3 tag ', () => {
        expect(wrapper.find('#board #moveHeader').length).toBe(1)
    })

    it(' h3 tag has Next Player: X text ', () => {
        console.log(wrapper.find('#board #moveHeader'))
        expect(wrapper.find('#board #moveHeader').text()).toBeTruthy()
    })

    it(' how <Squares /> are there? ', () => {
        expect(wrapper.find(Square).length).toBe(9)
    })

    it(' get the h3 Next Player : x text', () => {
        expect(wrapper.find('#moveHeader').length).toBe(1)
        expect(wrapper.find('#moveHeader').text()).toBe('Next Player:  X ')
    })
    

})






// snapshot testing suite 

describe(' Board matching snapshot test ', () => {
    it('testing board snpashot', () => {
        const tree = renderer.create(<Board 
        click={() => {}}
        winningSquares={() => {}}
        xIsNext={true}
        toggle={() => {}} 
        squares={multiDimTicTacScoreBoard} />).toJSON()

        expect(tree).toMatchSnapshot()
    })
})