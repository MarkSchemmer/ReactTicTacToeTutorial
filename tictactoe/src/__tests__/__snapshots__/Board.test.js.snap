// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` Board matching snapshot test  testing board snpashot 1`] = `
<div
  id="board"
  style={
    Object {
      "height": "600px",
      "margin": "auto",
      "width": "400px",
    }
  }
>
  <h3
    id="moveHeader"
  >
    Next Player: 
    <strong>
       
      X
       
    </strong>
  </h3>
  <div
    className="square"
    id={1}
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "",
        "border": "1px solid black",
        "display": "inline-block",
        "height": "100px",
        "margin": "0",
        "marginRight": "2px",
        "width": "110px",
      }
    }
  />
  <div
    className="square"
    id={2}
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "",
        "border": "1px solid black",
        "display": "inline-block",
        "height": "100px",
        "margin": "0",
        "marginRight": "2px",
        "width": "110px",
      }
    }
  />
  <div
    className="square"
    id={3}
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "",
        "border": "1px solid black",
        "display": "inline-block",
        "height": "100px",
        "margin": "0",
        "marginRight": "2px",
        "width": "110px",
      }
    }
  />
  <div
    className="square"
    id={4}
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "",
        "border": "1px solid black",
        "display": "inline-block",
        "height": "100px",
        "margin": "0",
        "marginRight": "2px",
        "width": "110px",
      }
    }
  />
  <div
    className="square"
    id={5}
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "",
        "border": "1px solid black",
        "display": "inline-block",
        "height": "100px",
        "margin": "0",
        "marginRight": "2px",
        "width": "110px",
      }
    }
  />
  <div
    className="square"
    id={6}
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "",
        "border": "1px solid black",
        "display": "inline-block",
        "height": "100px",
        "margin": "0",
        "marginRight": "2px",
        "width": "110px",
      }
    }
  />
  <div
    className="square"
    id={8}
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "",
        "border": "1px solid black",
        "display": "inline-block",
        "height": "100px",
        "margin": "0",
        "marginRight": "2px",
        "width": "110px",
      }
    }
  />
  <div
    className="square"
    id={9}
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "",
        "border": "1px solid black",
        "display": "inline-block",
        "height": "100px",
        "margin": "0",
        "marginRight": "2px",
        "width": "110px",
      }
    }
  />
  <div
    className="square"
    id={10}
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "",
        "border": "1px solid black",
        "display": "inline-block",
        "height": "100px",
        "margin": "0",
        "marginRight": "2px",
        "width": "110px",
      }
    }
  />
</div>
`;
